Cython/Compiler/ExprNodes.py

```diff
@@ -3931,12 +3974,13 @@ def analyse_as_pyobject(self, env, is_slice, getting, setting):
             if base_type in (list_type, tuple_type) and self.index.type.is_int:
                 item_type = infer_sequence_item_type(
                     env, self.base, self.index, seq_type=base_type)
-            if item_type is None:
-                item_type = py_object_type
-            self.type = item_type
-            if base_type in (list_type, tuple_type, dict_type):
-                # do the None check explicitly (not in a helper) to allow optimising it away
-                self.base = self.base.as_none_safe_node("'NoneType' object is not subscriptable")
+            if item_type is None or not item_type.is_pyobject:
```

Comment:
```suggestion
            if base_type in (list_type, tuple_type, dict_type):
                # do the None check explicitly (not in a helper) to allow optimising it away
                self.base = self.base.as_none_safe_node("'NoneType' object is not subscriptable")
            if item_type is None or not item_type.is_pyobject:
```

A good chunk of the test failures look fixable by reinstating these lines

Reply:
Just pinging this comment since I still think it fixes the worst of the remaining issues. (There may be a good reason why you're removed it that I'm missing of course)